### DISCUSSION TITLE
Issue with WI state change in PR (pull request)

### DISCUSSION THREAD

**Discussion by gubin-i-da** [tfsaggregator/tfsaggregator#379] [Q&A]
As described here
https://docs.microsoft.com/en-us/azure/devops/notifications/add-links-to-work-items?view=azure-devops#set-work-item-state-in-pull-request
work items can change state in handler of pull request.

Theoretically, one of the events to which the aggregator is subscribed should occur: creation, update, etc.
However, for some reason, the event is not processed by the aggregator.
Could you please explain why processing of WI state change in PR event handlers is not happening?

**Comment by jessehouwing** [comment]
That's a bug in Azure DevOps. It should trigger, but they (Azure DevOps) don't raise the event.

**Comment by gubin-i-da** [comment]
Hello!

1) I added a service that handles work item update event hooks. It simply writes the received JSON from TFS to a log file. And signed it for event handling.
2) After doing a pull request with associated work items: one task and one error, I get work item update events in the log file.

Working hypothesis: I think that the problem is not that the TFS aggregator is not receiving the event, but that the format of the received data may have changed.

To test the hypothesis in the attachment, cuts from the log file on the task and bug
[tfs_wi_update_hook_handler.zip](https://github.com/tfsaggregator/tfsaggregator/files/9394049/tfs_wi_update_hook_handler.zip)
.

**Reply by jessehouwing** [reply]
Can you share your bit of code? Whether a service hook triggers or not is unfortunately no guarantee an ISubscriber will receive the event or even the same event.

So please help us understand better what is happening and which apples you're comparing so what oranges.

**Reply by gubin-i-da** [reply]
https://github.com/gubin-i-da/tfs_hook_handlers
a little comment:
1) DA.TFS.WebHook.Handlers.sln - c# code with tfs hook handlers:
InspectController.cs - to inspect the JSON of any hook events
PRController.cs - experiment to run the tfs aggregator when pull request was ended
on launchSettings.json need change IP 192.168.192.161 to your current IP (ipconfig -all)
2) TFSAggregator2.ServerPlugin.policies - example of policies that worked when WI saving in it web UI but not worked when ending pull request with state "Closed" of related WI

**Reply by jessehouwing** [reply]
Can't access that repo. But it makes sense that the webhooks are still working.

TFS aggregator doesn't use the webhooks, it uses an in-process API on the server.

You could (should) switch over to the aggregator CLI. It's our new project that uses the webhooks and runs outhofhprocess of the TFS server.

**Reply by gubin-i-da** [reply]
Can I use the same policy settings as the tfs aggregator in a new project aggregator CLI?

**Reply by gubin-i-da** [reply]
>Can't access that repo.

I made a private repository and added you as a co-author, I thought this would be enough for access. 

Thank you!

**Reply by jessehouwing** [reply]
You'd have to rewrite the policy.

**Comment by gubin-i-da** [comment]
Jesse Hello!

I don't understand a few things in the description:

1) 
When Azure DevOps deploying locally, the project is uniquely identified by three parameters:
- dns of host where installed Azure DevOps Server
- collection name
- project name.

Here
https://tfsaggregator.github.io/docs/v3/setup/azure/
described only how to deploy a CLI to Azure (no local).

Here 
https://tfsaggregator.github.io/docs/v3/commands/map-commands/
described command map.local.rule and how to put in accordance to the project some rule.

How I need to define the project when I invoking the command? <azure_devops_host>/<collection>/<project> ?

2) When deployed locally, with aggregator CLI and Azure Devops Server on the same host (host1) and docker-part on a separate host with linux (host2), at what point in time does the aggregator CLI or Docker part know about Azure Devops Server ? 
![11111111111111111111111111](https://user-images.githubusercontent.com/110813893/186733922-6adb5774-527a-4af4-b82b-b71bd8acd0ef.png)
When configured by calling map.local.rule ? Or in some other settings for example
when configuring docker in the container_name and hostname parameters, as described here https://tfsaggregator.github.io/docs/v3/setup/docker/
"DNS name"?

May I ask you to describe how to deploy and configure the aggregator CLI for the described deployment configuration?

p.s.
I'm sorry, I edited your diagram for clarity.

**Comment by giuliov** [comment]
You define the Azure DevOps instance at [login](https://tfsaggregator.github.io/docs/v3/commands/authentication-commands/)). The URL should point to the Collection if I am not wrong.
You specify the project in the `map.local.rule` through the `--project` option.

BTW this discussion is in the wrong repo, as you are using the [CLI](https://github.com/tfsaggregator/aggregator-cli).

HTH

**Reply by gubin-i-da** [reply]
Giulio, thanks for the answer!

I moved [here](https://github.com/tfsaggregator/aggregator-cli/discussions/270)
